gf2-128/src/lib.rs

```diff
@@ -0,0 +1,207 @@
+//!
+//! ### A2M algorithm
+//! This is the other way round.
+//! Let `A` be an element of some finite field with `A = x * y`, where `x` is only known to Alice
+//! and `y` is only known to Bob. A is unknown to both parties and it is their goal that each of
+//! them ends up with a multiplicative share of A. So both parties start with `x` and `y` and want to
+//! end up with `a` and `b`, where `A = x + y = a * b`.
+//!
+//! This is an implementation for the extension field GF(2^128), which is a semi-honest adaptation
+//! of chapter 4 of <https://www.cs.umd.edu/~fenghao/paper/modexp.pdf>
+
+use rand::{Rng, SeedableRng};
+use rand_chacha::ChaCha12Rng;
+
+/// Masked values for an oblivious transfer
+pub struct MaskedPartialValue(pub [u128; 128], pub [u128; 128]);
+
+/// A multiplicative share of `A = a * b`
+pub struct MulShare(u128);
+
+impl MulShare {
+    /// Create a new `MulShare` holding a factor of `A`
+    pub fn new(share: u128) -> Self {
+        Self(share)
+    }
+
+    /// Return inner share
+    pub fn inner(&self) -> u128 {
+        self.0
+    }
+
+    /// Turn into an additive share and masked partial values
+    ///
+    /// This function returns
+    ///   * `AddShare` - The sender's additive share; this is `y` in the paper
+    ///   * `MaskedPartialValue` - Used for oblivious transfer; t0 and t1 in the paper
+    pub fn to_additive(&self) -> (AddShare, MaskedPartialValue) {
+        let mut rng = ChaCha12Rng::from_entropy();
+
+        let t0: [u128; 128] = std::array::from_fn(|_| rng.gen());
+        let t1: [u128; 128] = std::array::from_fn(|i| mul(self.inner(), 1 << i) ^ t0[i]);
+
+        let add_share = AddShare::new(t0.into_iter().fold(0, |acc, i| acc ^ i));
+        (add_share, MaskedPartialValue(t0, t1))
+    }
+
+    /// Create a multiplicative share from the output of an OT
+    ///
+    /// The `value` needs to be built by choices of an oblivious transfer
+    pub fn from_choice(value: [u128; 128]) -> Self {
+        Self::new(value.into_iter().fold(0, |acc, i| acc ^ i))
+    }
+}
+
+/// An additive share of `A = x + y`
+pub struct AddShare(u128);
+
+impl AddShare {
+    /// Create a new `AddShare` holding a summand of `A`
+    pub fn new(share: u128) -> Self {
+        Self(share)
+    }
+
+    /// Return inner share
+    pub fn inner(&self) -> u128 {
+        self.0
+    }
+
+    /// Turn into a multiplicative share and masked partial values
+    ///
+    /// This function returns
+    ///   * `MulShare` - The sender's multiplicative share
+    ///   * `MaskedPartialValue` - Used for oblivious transfer
+    pub fn to_multiplicative(&self) -> (MulShare, MaskedPartialValue) {
+        let mut rng = ChaCha12Rng::from_entropy();
+
+        let random: u128 = rng.gen();
+        let mut masks: [u128; 128] = std::array::from_fn(|_| rng.gen());
+        masks[127] = masks.into_iter().take(127).fold(0, |acc, i| acc ^ i);
+
+        let mul_share = MulShare::new(inverse(random));
+        let b0: [u128; 128] =
```

Comment:
I find that I have to pause and to think what `&` does here. I think should put a comment somewhere along the lines of 
// self.inner() & (1 << i) extracts self.inner()'s bit in position `i` (counting from the right) shifted left by `i`